sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py

```diff
@@ -0,0 +1,153 @@
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.testing import test_utils
+from apache_beam.testing.load_tests.load_test import LoadTest
+from apache_beam.testing.load_tests.load_test_metrics_utils import MeasureTime
+from apache_beam.testing.test_pipeline import TestPipeline
+
+PUBSUB_NAME = 'pubsub_read_performance'
+
+
+class PubsubReadPerfTest(LoadTest):
+
+  def __init__(self):
+    super(PubsubReadPerfTest, self).__init__()
+    self._setup_env()
+    self._setup_pubsub()
+    self._setup_pipeline()
+
+  def test(self):
+    (
+        self.pipeline
+        | 'Read from pubsub' >> ReadFromPubSub(
+            subscription=self.subscription.name,
+            with_attributes=True,
+            id_label='ack_id')
+        | 'Measure time' >> beam.ParDo(MeasureTime(self.metrics_namespace))
+    )
+
+  def cleanup(self):
+    test_utils.cleanup_subscriptions(
+      self.sub_client, [self.subscription, self.matcher_subscription])
+    test_utils.cleanup_topics(
+      self.pub_client, [self.topic])
+
+  def _setup_env(self):
+    if not self.pipeline.get_option('timeout'):
+      logging.error('--timeout argument is required.')
+      sys.exit(1)
+
+    self.num_of_messages = int(self.input_options.get('num_records'))
+    self.timeout = int(self.pipeline.get_option('timeout'))
+
+  def _setup_pubsub(self):
+    def init_input(number_of_messages):
+      for n in range(number_of_messages):
+        message = PubsubMessage(
+          bytes(str(n), 'utf-8'),
+          {'ack_id': bytes(str(n), 'utf-8')})
+        bytes_message = WriteToPubSub.to_proto_str(message)
+        self.pub_client.publish(self.topic.name, bytes_message)
+
+    from google.cloud import pubsub
+
+    self.uuid = str(uuid.uuid4())
+
+    self.pub_client = pubsub.PublisherClient()
+    self.topic = self.pub_client.create_topic(
+      self.pub_client.topic_path(self.project_id, PUBSUB_NAME + self.uuid)
+    )
+
+    self.sub_client = pubsub.SubscriberClient()
+    self.subscription = self.sub_client.create_subscription(
+      self.sub_client.subscription_path(self.project_id, PUBSUB_NAME + self.uuid),
+      self.topic.name
+    )
+    self.matcher_subscription = self.sub_client.create_subscription(
+      self.sub_client.subscription_path(self.project_id, PUBSUB_NAME + '_matcher_' + self.uuid),
+      self.topic.name
+    )
+
+    init_input(self.num_of_messages)
+
+
+  def _setup_pipeline(self):
+    pubsub_msg_verifier = PubSubMessageMatcher(
+        self.project_id,
+        self.matcher_subscription.name,
+        expected_msg_len=self.num_of_messages,
+        timeout=self.timeout
+    )
+
+    self.extra_opts = {
+        'on_success_matcher': all_of(pubsub_msg_verifier),
+        'wait_until_finish_duration': self.timeout * 1000,
+        'streaming': True,
+        'save_main_session': True
+    }
+
+    args = self.pipeline.get_full_options_as_args(**self.extra_opts)
```

Comment:
I think since we don't use `self.extra_opts` anywhere else, the assignment to `self` is redundant.